fl4health/strategies/feddg_ga_strategy.py

```diff
@@ -0,0 +1,382 @@
+
+        Returns:
+            (Tuple[Optional[Parameters], Dict[str, Scalar]]) A tuple containing the aggregated parameters
+                and the aggregated fit metrics.
+        """
+
+        _, metrics_aggregated = super().aggregate_fit(server_round, results, failures)
+
+        self.train_metrics = {}
+        for result in results:
+            self.train_metrics[result[0].cid] = result[1].metrics
+
+        parameters_aggregated = ndarrays_to_parameters(self.weight_and_aggregate_results(results))
+
+        return parameters_aggregated, metrics_aggregated
+
+    def aggregate_evaluate(
+        self,
+        server_round: int,
+        results: List[Tuple[ClientProxy, EvaluateRes]],
+        failures: List[Union[Tuple[ClientProxy, EvaluateRes], BaseException]],
+    ) -> Tuple[Optional[float], Dict[str, Scalar]]:
+        """
+        Aggregate evaluation losses using weighted average.
+
+        Collects the evaluation metrics and update the adjustment weights, which will be used
+        when aggregating the results for the next round.
+
+        Args:
+            server_round: (int) the current server round.
+            results: (List[Tuple[ClientProxy, FitRes]]) The clients' evaluate results.
+            failures: (List[Union[Tuple[ClientProxy, FitRes], BaseException]]) the clients' evaluate failures.
+
+        Returns:
+            (Tuple[Optional[float], Dict[str, Scalar]]) A tuple containing the aggregated evaluation loss
+                and the aggregated evaluation metrics.
+        """
+        loss_aggregated, metrics_aggregated = super().aggregate_evaluate(server_round, results, failures)
+
+        self.evaluation_metrics = {}
+        for result in results:
+            cid = result[0].cid
+            self.evaluation_metrics[cid] = result[1].metrics
+            # adding the loss to the metrics
+            val_loss_key = FairnessMetricType.LOSS.value
+            self.evaluation_metrics[cid][val_loss_key] = loss_aggregated  # type: ignore
+
+        # Updating the weights at the end of the training round
+        cids = [result[0].cid for result in results]
+        self.update_weights_by_ga(server_round, cids)
+
+        return loss_aggregated, metrics_aggregated
+
+    def weight_and_aggregate_results(self, results: List[Tuple[ClientProxy, FitRes]]) -> NDArrays:
+        """
+        Aggregate results by weighing them against the adjustment weights and then summing them.
+
+        Args:
+            results: (List[Tuple[ClientProxy, FitRes]]) The clients' fit results.
+
+        Returns:
+            (NDArrays) the weighted and aggregated results.
+        """
+
+        aggregated_results: Optional[NDArrays] = None
+        for result in results:
+            cid = result[0].cid
+
+            # initializing adjustment weights for this client if they don't exist yet
+            if cid not in self.adjustment_weights:
+                self.adjustment_weights[cid] = INITIAL_ADJUSTMENT_WEIGHT
```

Comment:
It we don't have an adjustment weight for a client (i.e. it joins the fitting round later on or something), then the sum of the adjustment weights will no longer be 1.0. If we add in a client later on, should we not normalize the weight adjustment coefficients to preserve this property?

Reply:
Yes, Fatemeh commented on it, an overlook on my part. I'm changing it to 1/number-of-clients.

Reply:
About adding a client, I am not sure what to do. Right now, it's easier in the code to recalculate the weight adjustment at the beginning of every fit_round and maybe that is a good behaviour unless the amount of clients changes dramatically between rounds.

Reply:
Yeah I think that's fine. In practice, at least based on my reading, this method isn't really well defined for a situation where you are adding and losing clients in each round to be honest.